Reloaded.Imgui.Hook.OpenGL/ImguiHookOpenGL.cs

```diff
@@ -0,0 +1,221 @@
+            if (_wglSwapBuffersRecursionLock)
+            {
+                Debug.WriteLine($"[OpenGL wglSwapBuffers] Discarding via Recursion Lock");
+                return _wglSwapBuffersHook.OriginalFunction.Value.Invoke(hdc);
+            }
+
+            _wglSwapBuffersRecursionLock = true;
+            try
+            {
+                return SwapBuffersCommon(hdc, () => _wglSwapBuffersHook.OriginalFunction.Value.Invoke(hdc), "wglSwapBuffers");
+            }
+            finally
+            {
+                _wglSwapBuffersRecursionLock = false;
+            }
+        }
+
+        private int SwapBuffersImpl(IntPtr hdc)
+        {
+            // With multi-viewports, ImGui might call SwapBuffers again; so we need to prevent stack overflow here.
+            if (_swapBuffersRecursionLock)
+            {
+                Debug.WriteLine($"[OpenGL SwapBuffers] Discarding via Recursion Lock");
+                return _swapBuffersHook.OriginalFunction.Value.Invoke(hdc);
+            }
+
+            _swapBuffersRecursionLock = true;
+            try
+            {
+                return SwapBuffersCommon(hdc, () => _swapBuffersHook.OriginalFunction.Value.Invoke(hdc), "SwapBuffers");
+            }
+            finally
+            {
+                _swapBuffersRecursionLock = false;
+            }
+        }
+
+        private int SwapBuffersCommon(IntPtr hdc, Func<int> originalFunction, string hookName)
+        {
+            // Get window handle from device context
+            var windowHandle = WindowFromDC(hdc);
+
+            // Ignore windows which don't belong to us.
+            if (!ImguiHook.CheckWindowHandle(windowHandle))
+            {
+                Debug.WriteLine($"[OpenGL {hookName}] Discarding Window Handle {(long)windowHandle:X}");
+                return originalFunction();
+            }
+
+            // Prevent double rendering if an app uses both methods or one calls the other.
+            if (_isRendering)
+                return originalFunction();
+
+            _isRendering = true;
+            try
+            {
+                if (!_initialized)
+                {
+                    _deviceContext = hdc;
+                    _windowHandle = windowHandle;
+                    if (_windowHandle == IntPtr.Zero)
+                        return originalFunction();
+
+                    Debug.WriteLine($"[OpenGL {hookName}] Init, Window Handle {(long)windowHandle:X}");
+                    ImguiHook.InitializeWithHandle(windowHandle);
+
+                    // Initialize OpenGL3 backend with GLSL version string
+                    // Using "#version 130" for OpenGL 3.0+ compatibility
+                    ImGui.ImGuiImplOpenGL3Init("#version 130");
+                    _initialized = true;
+                }
+
+                // Log which method is being used (once per method)
+                if (hookName == "wglSwapBuffers" && !_wglUsed)
```

Comment:
Avoid the string comparison if possible.

Remember, this is being done on the render thread, and thus directly has perf impact.
Use an enum here, as parameter for the type used, which the JIT could compile away (if it feels like inlining).
But I don't think that's even needed here at all (see below).

The `_gdiUsed` & `_wglUsed` don't appear to be used in any logic, and thus are redundant. Presumably leftovers from earlier code. Have the likes of `Debug.WriteLine($"[OpenGL] Application is using wglSwapBuffers");` in the caller, if a print is needed.

It appears that the original function is called on return in all cases. So you can skip passing the `func` entirely, and just call it from the caller after running `SwapBuffersCommon`. That alongside the log.

Reply:
Addressed! :)